api/desecapi/serializers.py

```diff
@@ -218,12 +129,152 @@ def validate_type(value):
+                ret.append(validated)
+                errors.append({})
+
+        self.partial = partial
+
+        if any(errors):
+            raise ValidationError(errors)
+
+        return ret
+
+    def create(self, validated_data):
+        index = {(data.get('subname', ''), data['type']): data for data in validated_data}
+        for (subname, type_) in index:
+            if not index[(subname, type_)].get('type', None):
+                raise ValidationError('type is required')
+            if not index[(subname, type_)].get('ttl', None):
+                raise ValidationError('TTL is required')
+        if len(index) != len(validated_data):
+            raise ValidationError()  # TODO where to put this? Check non_field_errors elsewhere in DRF
+        return super().create(validated_data)
+
+    def update(self, instance, validated_data):
+        instance_index = {(rrset.subname, rrset.type): rrset for rrset in instance}
+        data_index = {(data.get('subname', ''), data['type']): data for data in validated_data}
+
+        created = data_index.keys() - instance_index.keys()
+        updated = {(d.get('subname', ''), d['type']) for d in validated_data if d.get('records', None) != []} - created
+        deleted = {(d.get('subname', ''), d['type']) for d in validated_data if d.get('records', None) == []}
+
+        # TODO consider moving those validation checks to a more appropriate location
+        if len(validated_data) != len(data_index.keys()):
+            raise ValidationError()  # TODO where to put this? Check non_field_errors elsewhere in DRF
+                                     #  possibly to_internal_value()
+
+        for subname, type_ in created:
+            if not data_index[(subname, type_)].get('type', None):
+                raise ValidationError('type is required')
+            if not data_index[(subname, type_)].get('ttl', None):
+                raise ValidationError('TTL is required')
+            if not data_index[(subname, type_)].get('records', []):
+                raise ValidationError('New resource record set with subname %s and type %s does not have records.' %
+                                      (subname, type_))
+
+        assert created | updated | deleted == data_index.keys(), 'Given data not covered. Expected %s, but got %s ' \
+                                                                 'and %s and %s.' % (data_index.keys(), created,
+                                                                                     updated, deleted)
+        assert not created & updated and not updated & deleted and not deleted & created, \
+            'Created, updated, deleted not pairwise disjoint.'
+        assert data_index.keys() | instance_index.keys() == data_index.keys(), \
+            'Instances is not a subset of data.'
+
+        # Note that for the following create and update actions, the data passed along is treated as validated,
+        # no further validation is conducted!
```

Comment:
That sounds like that would be a problem, as data is *treated* as validated. Isn't it also *actually* validated? (`data_index` comes from `validated_data`.)

Reply:
This was a personal note to not forget that I was moving away checks from the `create` and `update` methods to elsewhere in the serializer. This comment was supposed to remind me to see if the checks elsewhere are actually effective.